File: problems/chapter11/mshibatatt/002.py
# https://atcoder.jp/contests/abc120/tasks/abc120_d

class UnionFind():
    def __init__(self, n) -> None:
        self.par = [-1] * n
        self.siz = [1] * n

    def root(self, x):
        if self.par[x] == -1:
            return x
        else:
            self.par[x] = self.root(self.par[x])
            return self.par[x]
    
    def issame(self, x, y):
        return self.root(x) == self.root(y)
    
    def unite(self, x, y):
        x = self.root(x)
        y = self.root(y)
        if x == y:
            return False
        
        # make the number of child of y is smaller than that of x 
        if self.siz[x] < self.siz[y]:
            x, y = y, x
        # make y child of x
        self.par[y] = x
        self.siz[x] += self.siz[y]
        
        return True

    def size(self, x):
        return self.siz[self.root(x)]


def main():
    N, M = map(int, input().split())
    edges = []
    for _ in range(M):
        a, b = map(int, input().split())
        edges.append([a-1, b-1])

    uf = UnionFind(N)
    # inconv after all bridge has broken
    output = [int(N*(N-1)/2)]

    # after 0~i bridges have broken
    #   <=> after constructing i+1~M bridges from no bridge 
    for _ in range(M-1):
        bridge = edges.pop()
        # count inconv island by substructing the number of connected islands
        if uf.issame(bridge[0], bridge[1]):
            output.append(output[-1])
        else:
            output.append(output[-1] - uf.size(bridge[0])*uf.size(bridge[1]))
        uf.unite(bridge[0], bridge[1])

    # print answer
    for i in output[::-1]:
        print(i)

if __name__=='__main__':
    main()

File: problems/chapter11/miiitomi/002.cpp
// AtCoder ABC120   D - Decayed Bridges
// https://atcoder.jp/contests/abc120/tasks/abc120_d

#include <iostream>
#include <vector>
using namespace std;

struct UnionFind {
    vector<int> par, siz;

    UnionFind(int n) : par(n, -1), siz(n, 1) {}

    int root(int x) {
        if (par[x] == -1) return x;
        else return par[x] = root(par[x]);
    }

    bool issame(int x, int y) {
        return root(x) == root(y);
    }

    bool unite(int x, int y) {
        x = root(x);
        y = root(y);
        if (x == y) return false;
        if (siz[x] < siz[y]) swap(x, y);
        par[y] = x;
        siz[x] += siz[y];
        return true;
    }

    int size(int x) {
        return siz[root(x)];
    }
};

int main() {
    int N, M;
    cin >> N >> M;
    vector<int> A(M), B(M);
    for (int i = 0; i < M; i++) {
        cin >> A[i] >> B[i];
        A[i]--;
        B[i]--;
    }
    vector<int64_t> ans(M);
    ans[M-1] = ((int64_t)N * (N - 1)) / 2LL;

    UnionFind uf(N);

    for (int j = M - 1; j > 0; j--) {
        ans[j-1] = ans[j];
        int a_size = uf.size(A[j]);
        int b_size = uf.size(B[j]);
        bool united = uf.unite(A[j], B[j]);
        if (united) {
            ans[j-1] -= (int64_t)a_size * b_size;
        }
    }

    for (int i = 0; i < M; i++) {
        cout << ans[i] << endl;
    }
}
